### ISSUE TITLE
Roundabouts vs Signals

### ISSUE THREAD

**Issue by bahrus** [proposal-signals/proposal-signals#127]
I favor this proposal.  I've used knockout extensively, and I think this does solve some problems well, and is a step forward.  However, I don't think I will rely exclusively, if at all, on signals in my own custom elements, as it leaves me wanting for something better.  That thing that I think is better is below.  I think it is too early to make a proposal out of it, but I hope it will generate some thought and some competing ideas.

Don't worry, I plan to close this issue soon.

# Roundabouts vs Signals

The world needs both traffic signals and roundabouts.  This shouldn't be an either or.

Signals:

```JavaScript
const counter = new Signal.State(0);
const isEven = new Signal.Computed(() => (counter.get() & 1) == 0);
const parity = new Signal.Computed(() => isEven.get() ? "even" : "odd");

effect(() => element.innerText = parity.get());

// Simulate external updates to counter...
setInterval(() => counter.set(counter.get() + 1), 1000);
```

Roundabouts:

<img src="https://www.trafficdepot.ca/wp-content/uploads/2020/08/reg6bb.png" height="250px">

```JavaScript
const checkIfEven = ({counter}) => ({isEven: counter & 1 === 0});
const determineParity = ({isEven}) => ({parity: isEven ? 'even' : 'odd'});
const setInnerText = ({parity}) => ({'?.element?.innerText': parity};
const [vm, propagator] = await roundabout(
    {element, checkIfEven, determineParity, innerText}, 
    {   
        propagate: {count: 0},
        do_checkIfEven_on: 'count', do_determineParity_on: 'isEven', do_setInnerText_on: 'parity'
    }
);

setInterval(() => vm.count++, 1000);
```

Same number of statements, but for roundabout, one of the statements is admittedly long.

All the functions are side effect free and don't do any state mutation at all!  Purely functional.

roundabout can JSON serialize one of the arguments, making parsing the instructions easier on the browser.

roundabout does require manually figuring out the dependencies ("sources"), but the silver lining is it gives the user more transparent power, especially as they can direct traffic based on truthy conditions.  For example:

```JavaScript
const [vm, propagator] = await roundabout(
    {element, checkIfEven, determineParity, setInnerText}, 
    {   
        propagate: {count: 0},
        do_checkIfEven_on: 'count', do_determineParity_on: 'isEven',
        setInnerText: {
            ifAllOf: ['count', 'isEven', 'parity']
        }
    }
);
```

That manual calculation could be done during compile time.

Less run time analysis?

Fewer (nested) parenthesis.

State is all in one place -- the vm, which could also be the custom element class.

Lower learning curve?

roundabout also doesn't execute code if the field value is unchanged.

No pub/sub required!

No creation of getters/setters required (other than count)!

Basically, what round about does is looks at what subset of properties of the view model is returned from the action methods (checkIfEven, determineParity, setInnerText), and directs traffic accordingly after doing an Object.assignGingerly.

propagator is an EventTarget, that publishes events when the propagate properties are changed (just count).

What standard would help?

1.  Object.assignGingerly.

2.  Being able to publish the dependencies in a uniform way:

```JavaScript
const parity = ({isEven}) => ({parity: isEven ? 'even' : 'odd'});
parity.do = {
    on: 'isEven'.
    if: [],
    ifOneOf:[],
    ifEquals:[],
    etc.
}
```

which must be JSON serializable.

3.  Reducing the footprint

Maybe this could be done with decorators

**Comment by littledan** [comment]
Every proposal needs to come with a logo from now on! Well done, sir. You have done this correctly.

More seriously: ergonomics of use of signals is very important, and I'm glad that you're experimenting here. We're focusing initially on the reactivity core, and a major near-term goal is to ensure that this system is sufficiently "expressive" such that other systems, like your roundabouts, could be implemented on top of an underlying signal graph. Do you think that approach would make sense for you?
